Got the failure locally. It's a pretty slow test 😅 
```
alex@alangford build % ./bin/lldb-dotest -p TestWriteMemoryWithHWBreakpoint.py -t
/Applications/Xcode.app/Contents/Developer/usr/bin/python3 /Users/alex/Projects/OSS/llvm-project/lldb/test/API/dotest.py --triple arm64-apple-darwin27.0.0 -u CXXFLAGS -u CFLAGS --build-dir /Users/alex/Projects/OSS/llvm-project/build/lldb-test-build.noindex --executable /Users/alex/Projects/OSS/llvm-project/build/./bin/lldb --compiler /Users/alex/Projects/OSS/llvm-project/build/./bin/clang --dsymutil /Users/alex/Projects/OSS/llvm-project/build/./bin/dsymutil --make /usr/bin/make --lldb-libs-dir /Users/alex/Projects/OSS/llvm-project/build/./lib --llvm-tools-dir /Users/alex/Projects/OSS/llvm-project/build/./bin --libcxx-include-dir /Users/alex/Projects/OSS/llvm-project/build/include/c++/v1 --libcxx-library-dir /Users/alex/Projects/OSS/llvm-project/build/lib --framework /Users/alex/Projects/OSS/llvm-project/build/bin/LLDB.framework --arm64e-debugserver --lldb-obj-root /Users/alex/Projects/OSS/llvm-project/build/tools/lldb --cmake-build-type RelWithDebInfo -p TestWriteMemoryWithHWBreakpoint.py -t
Skipping the following test categories: libstdcxx, msvcstl, dwo, pdb, llgs, fork
Change dir to: /Users/alex/Projects/OSS/llvm-project/lldb/test/API/functionalities/breakpoint/hardware_breakpoints/write_memory_with_hw_breakpoint
runCmd: settings clear --all

output:

runCmd: settings set symbols.enable-external-lookup false

output:

runCmd: settings set target.inherit-tcc true

output:

runCmd: settings set target.disable-aslr false

output:

runCmd: settings set target.detach-on-error false

output:

runCmd: settings set target.auto-apply-fixits false

output:

runCmd: settings set plugin.process.gdb-remote.packet-timeout 60

output:

runCmd: settings set symbols.clang-modules-cache-path "/Users/alex/Projects/OSS/llvm-project/build/lldb-test-build.noindex/module-cache-lldb"

output:

runCmd: settings set use-color false

output:

runCmd: settings set show-statusline false

output:

runCmd: settings set target.check-vo-ownership true

output:

/usr/bin/make VPATH=/Users/alex/Projects/OSS/llvm-project/lldb/test/API/functionalities/breakpoint/hardware_breakpoints/write_memory_with_hw_breakpoint -C /Users/alex/Projects/OSS/llvm-project/build/lldb-test-build.noindex/functionalities/breakpoint/hardware_breakpoints/write_memory_with_hw_breakpoint/TestWriteMemoryWithHWBreakpoint -I /Users/alex/Projects/OSS/llvm-project/lldb/test/API/functionalities/breakpoint/hardware_breakpoints/write_memory_with_hw_breakpoint -I /Users/alex/Projects/OSS/llvm-project/lldb/packages/Python/lldbsuite/test/make -f /Users/alex/Projects/OSS/llvm-project/lldb/test/API/functionalities/breakpoint/hardware_breakpoints/write_memory_with_hw_breakpoint/Makefile all TRIPLE=arm64-apple-darwin27.0.0 CC=/Users/alex/Projects/OSS/llvm-project/build/bin/clang CC_TYPE=clang CXX=/Users/alex/Projects/OSS/llvm-project/build/bin/clang++ LLVM_AR=/Users/alex/Projects/OSS/llvm-project/build/./bin/llvm-ar AR=/Users/alex/Projects/OSS/llvm-project/build/./bin/llvm-ar OBJCOPY=/Users/alex/Projects/OSS/llvm-project/build/./bin/llvm-objcopy STRIP=/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/strip ARCHIVER=/Users/alex/Projects/OSS/llvm-project/build/./bin/llvm-ar DWP=/Users/alex/Projects/OSS/llvm-project/build/./bin/llvm-dwp AR=libtool DSYMUTIL=/Users/alex/Projects/OSS/llvm-project/build/./bin/dsymutil 'CODESIGN=codesign --entitlements /Users/alex/Projects/OSS/llvm-project/lldb/packages/Python/lldbsuite/test/make/entitlements-macos.plist' CLANG_MODULE_CACHE_DIR=/Users/alex/Projects/OSS/llvm-project/build/lldb-test-build.noindex/module-cache-clang LIBCPP_INCLUDE_DIR=/Users/alex/Projects/OSS/llvm-project/build/include/c++/v1 LIBCPP_LIBRARY_DIR=/Users/alex/Projects/OSS/llvm-project/build/lib LLDB_OBJ_ROOT=/Users/alex/Projects/OSS/llvm-project/build/tools/lldb OS=Darwin HOST_OS=Darwin

make: Nothing to be done for `all'.


runCmd: file /Users/alex/Projects/OSS/llvm-project/build/lldb-test-build.noindex/functionalities/breakpoint/hardware_breakpoints/write_memory_with_hw_breakpoint/TestWriteMemoryWithHWBreakpoint/a.out

output: Current executable set to '/Users/alex/Projects/OSS/llvm-project/build/lldb-test-build.noindex/functionalities/breakpoint/hardware_breakpoints/write_memory_with_hw_breakpoint/TestWriteMemoryWithHWBreakpoint/a.out' (arm64).


runCmd: breakpoint set -b main --hardware

output: Breakpoint 1: where = a.out`main + 24 at main.cpp:8:28, address = 0x00000001000004ac


runCmd: run

output: Process 5148 launched: '/Users/alex/Projects/OSS/llvm-project/build/lldb-test-build.noindex/functionalities/breakpoint/hardware_breakpoints/write_memory_with_hw_breakpoint/TestWriteMemoryWithHWBreakpoint/a.out' (arm64)
Process 5148 stopped
* thread #1, queue = 'com.apple.main-thread', stop reason = breakpoint 1.1
       frame #0: 0x00000001043004ac a.out`main(argc=1, argv=0x000000016baff270) at main.cpp:8:28
   5    }
   6
   7    int main (int argc, char const *argv[]) {
-> 8      return hw_break_function(argc) ? 0 : 1;
                                   ^
   9    }


/usr/bin/make VPATH=/Users/alex/Projects/OSS/llvm-project/lldb/test/API/functionalities/breakpoint/hardware_breakpoints/write_memory_with_hw_breakpoint -C /Users/alex/Projects/OSS/llvm-project/build/lldb-test-build.noindex/functionalities/breakpoint/hardware_breakpoints/write_memory_with_hw_breakpoint/TestWriteMemoryWithHWBreakpoint -I /Users/alex/Projects/OSS/llvm-project/lldb/test/API/functionalities/breakpoint/hardware_breakpoints/write_memory_with_hw_breakpoint -I /Users/alex/Projects/OSS/llvm-project/lldb/packages/Python/lldbsuite/test/make -f /Users/alex/Projects/OSS/llvm-project/lldb/test/API/functionalities/breakpoint/hardware_breakpoints/write_memory_with_hw_breakpoint/Makefile all TRIPLE=arm64-apple-darwin27.0.0 CC=/Users/alex/Projects/OSS/llvm-project/build/bin/clang CC_TYPE=clang CXX=/Users/alex/Projects/OSS/llvm-project/build/bin/clang++ LLVM_AR=/Users/alex/Projects/OSS/llvm-project/build/./bin/llvm-ar AR=/Users/alex/Projects/OSS/llvm-project/build/./bin/llvm-ar OBJCOPY=/Users/alex/Projects/OSS/llvm-project/build/./bin/llvm-objcopy STRIP=/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/strip ARCHIVER=/Users/alex/Projects/OSS/llvm-project/build/./bin/llvm-ar DWP=/Users/alex/Projects/OSS/llvm-project/build/./bin/llvm-dwp AR=libtool DSYMUTIL=/Users/alex/Projects/OSS/llvm-project/build/./bin/dsymutil 'CODESIGN=codesign --entitlements /Users/alex/Projects/OSS/llvm-project/lldb/packages/Python/lldbsuite/test/make/entitlements-macos.plist' CLANG_MODULE_CACHE_DIR=/Users/alex/Projects/OSS/llvm-project/build/lldb-test-build.noindex/module-cache-clang LIBCPP_INCLUDE_DIR=/Users/alex/Projects/OSS/llvm-project/build/include/c++/v1 LIBCPP_LIBRARY_DIR=/Users/alex/Projects/OSS/llvm-project/build/lib LLDB_OBJ_ROOT=/Users/alex/Projects/OSS/llvm-project/build/tools/lldb OS=Darwin HOST_OS=Darwin

make: Nothing to be done for `all'.


runCmd: process launch --stop-at-entry

output: Process 5155 launched: '/Users/alex/Projects/OSS/llvm-project/build/lldb-test-build.noindex/functionalities/breakpoint/hardware_breakpoints/write_memory_with_hw_breakpoint/TestWriteMemoryWithHWBreakpoint/a.out' (arm64)


Ran command:
"process launch --stop-at-entry"

Got output:
Process 5155 launched: '/Users/alex/Projects/OSS/llvm-project/build/lldb-test-build.noindex/functionalities/breakpoint/hardware_breakpoints/write_memory_with_hw_breakpoint/TestWriteMemoryWithHWBreakpoint/a.out' (arm64)

Expecting regex pattern: "Process .* launched: .*a.out" (was found, matched "Process 5155 launched: '/Users/alex/Projects/OSS/llvm-project/build/lldb-test-build.noindex/functionalities/breakpoint/hardware_breakpoints/write_memory_with_hw_breakpoint/TestWriteMemoryWithHWBreakpoint/a.out")

runCmd: breakpoint add name --hardware -- 'hw_break_function'

output: Breakpoint 1: where = a.out`hw_break_function(int) + 8 at main.cpp:4:10, address = 0x0000000102fd4478


Expecting pattern: ^Breakpoint (?P<bpno>[0-9]+): (?P<num_locations>[0-9]+) locations\.$
Not matched

Expecting pattern: ^Breakpoint (?P<bpno>[0-9]+): (?P<num_locations>no) locations \(pending\)\.
Not matched

Expecting pattern: ^Breakpoint (?P<bpno>[0-9]+): where = (?P<module>.*)`(?P<symbol>[+\-]{0,1}[^+]+)( \+ (?P<offset>[0-9]+)){0,1}( \[inlined\] (?P<inline_symbol>.*)){0,1} at (?P<file>[^:]+):(?P<line_no>[0-9]+)(?P<column>(:[0-9]+)?), address = (?P<address>0x[0-9a-fA-F]+)$
Matched

<bound method SBProcess.Kill of SBProcess: pid = 5148, state = exited, threads = 1, executable = a.out>: success

<bound method SBProcess.Kill of SBProcess: pid = 5155, state = exited, threads = 1, executable = a.out>: success

FAIL: LLDB (/Users/alex/Projects/OSS/llvm-project/build/bin/clang-arm64) :: test_copy_memory_with_hw_break (TestWriteMemoryWithHWBreakpoint.WriteMemoryWithHWBreakpoint)
Log Files:
 - /Users/alex/Projects/OSS/llvm-project/build/lldb-test-build.noindex/functionalities/breakpoint/hardware_breakpoints/write_memory_with_hw_breakpoint/TestWriteMemoryWithHWBreakpoint/Failure_test_copy_memory_with_hw_break.log
Restore dir to: /Users/alex/Projects/OSS/llvm-project/build
======================================================================
FAIL: test_copy_memory_with_hw_break (TestWriteMemoryWithHWBreakpoint.WriteMemoryWithHWBreakpoint)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/alex/Projects/OSS/llvm-project/lldb/packages/Python/lldbsuite/test/decorators.py", line 220, in wrapper
    return func(*args, **kwargs)
  File "/Users/alex/Projects/OSS/llvm-project/lldb/test/API/functionalities/breakpoint/hardware_breakpoints/write_memory_with_hw_breakpoint/TestWriteMemoryWithHWBreakpoint.py", line 48, in test_copy_memory_with_hw_break
    self.assertTrue(error.Success())
AssertionError: False is not true
Config=arm64-/Users/alex/Projects/OSS/llvm-project/build/bin/clang
----------------------------------------------------------------------
Ran 1 test in 242.477s

FAILED (failures=1)
```

--- a/lldb/test/API/functionalities/breakpoint/hardware_breakpoints/write_memory_with_hw_breakpoint/TestWriteMemoryWithHWBreakpoint.py
+++ b/lldb/test/API/functionalities/breakpoint/hardware_breakpoints/write_memory_with_hw_breakpoint/TestWriteMemoryWithHWBreakpoint.py
@@ -46,4 +46,4 @@ class WriteMemoryWithHWBreakpoint(HardwareBreakpointTestBase):
 
         result = process.WriteMemory(address, data, error)
         self.assertTrue(error.Success())
-        self.assertEquals(result, len(data))
+        self.assertEqual(result, len(data))

--- a/lldb/test/API/functionalities/breakpoint/hardware_breakpoints/write_memory_with_hw_breakpoint/TestWriteMemoryWithHWBreakpoint.py
+++ b/lldb/test/API/functionalities/breakpoint/hardware_breakpoints/write_memory_with_hw_breakpoint/TestWriteMemoryWithHWBreakpoint.py
@@ -14,7 +14,7 @@ from functionalities.breakpoint.hardware_breakpoints.base import *
 class WriteMemoryWithHWBreakpoint(HardwareBreakpointTestBase):
 
     @skipTestIfFn(HardwareBreakpointTestBase.hw_breakpoints_unsupported)
-    def test_copy_memory_with_hw_break(self):
+    def test_write_memory_with_hw_break(self):
         self.build()
         exe = self.getBuildArtifact("a.out")
 

--- a/lldb/tools/debugserver/source/DNBBreakpoint.cpp
+++ b/lldb/tools/debugserver/source/DNBBreakpoint.cpp
@@ -135,7 +135,7 @@ size_t DNBBreakpointList::FindBreakpointsThatOverlapRange(
       iterator prev_pos = pos;
       --prev_pos;
       if (prev_pos->second.IntersectsRange(addr, size, NULL, NULL, NULL))
-        bps.push_back(&pos->second);
+        bps.push_back(&prev_pos->second);
     }
 
     while (pos != end) {
@@ -147,10 +147,10 @@ size_t DNBBreakpointList::FindBreakpointsThatOverlapRange(
         break;
 
       // Check if this breakpoint overlaps, and if it does, add it to the list
-      if (pos->second.IntersectsRange(addr, size, NULL, NULL, NULL)) {
+      if (pos->second.IntersectsRange(addr, size, NULL, NULL, NULL))
         bps.push_back(&pos->second);
-        ++pos;
-      }
+
+      ++pos;
     }
   }
   return bps.size();

